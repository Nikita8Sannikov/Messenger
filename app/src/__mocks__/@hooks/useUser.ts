import { mockUserService } from "./useProfile";

const useUser = jest.fn().mockReturnValue(mockUserService);

export default useUser; 