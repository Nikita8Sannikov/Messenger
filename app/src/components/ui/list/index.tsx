import { memo, type ReactNode } from "react";
import List, { type ListProps } from "@mui/material/List";

interface IListComponent extends ListProps {
    className?: string;
    children: ReactNode;
};

// Базовый компонент списка
export default memo(function ListComponent({ className, children, ...props }: IListComponent) {
	return <List className={className} {...props}>
		{children}
	</List>;
});