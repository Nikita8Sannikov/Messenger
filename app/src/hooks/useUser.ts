import useProfile from "@hooks/useProfile";
import { MY_ID } from "@utils/constants";

// Возврат сущности "Пользователь" по переданному идентификатору
export default function useUser(userId: string = MY_ID) {
    const profile = useProfile(userId);
    return profile.user;
}