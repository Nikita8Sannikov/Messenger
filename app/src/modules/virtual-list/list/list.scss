.virtual-list {
    &__loading {
        width: 100%;
        height: max-content;
        margin: 0 auto;
        position: relative;
    }

    &__no-content {
        display: flex;
        justify-content: center;
        align-items: center;
        height: inherit;
        width: inherit;
    }
}