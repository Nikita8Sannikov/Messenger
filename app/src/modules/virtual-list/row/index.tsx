import { type CSSProperties, memo, type ReactElement, type ReactNode } from "react";

import SpinnerComponent from "@components/ui/spinner";

import "./row.scss";

interface IListRow<T> {
    index: number;
    style: CSSProperties;
    data: {
        items: T[];
        itemWidth: number;
        hasMore: boolean;
    };
    options: {
        cols: number;
        gap: number;
        itemHeight: number;
    };
    renderCb: (props: { item: T, itemWidth: number, itemHeight: number; }) => ReactNode;
};

/**
 * Элемент, который отрисовывается в строке виртуального списка. 
 * Мемоизирован, так что перерисовывается только при изменении props.
 * Из родителььского компонента передаются стили для размещения элемента в строке.
 * Количество элементов в строке определяется в месте использования компонента виртуального списка.
 */
export default memo(function ListRow<T extends { id: string }>({ index, style, data, options, renderCb }: IListRow<T>) {
	const { cols, gap, itemHeight } = options;
	const { items, itemWidth, hasMore } = data;
	const start = index * cols;
	const rowItems = items.slice(start, start + cols);

	// Строка со спиннером загрузки (идет последняя в списке)
	if (!rowItems.length && hasMore) {
		return <div className="list__loading-row" style={style}>
			<SpinnerComponent />
		</div>;
	}

	// Обычная строка с картинками
	return <div className="list__row" style={{ ...style, gap }}>
		{rowItems.map(item => renderCb({ item, itemWidth, itemHeight }))}
	</div>;
}) as <T extends { id: string; }>(props: IListRow<T>) => ReactElement;