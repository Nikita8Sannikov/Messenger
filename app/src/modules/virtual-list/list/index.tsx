import type { ReactElement, ReactNode, Ref } from "react";
import { forwardRef, useCallback, useEffect, useImperativeHandle, useMemo, useRef } from "react";
import { FixedSizeList as List, type ListOnScrollProps } from "react-window";
import InfiniteLoader from "react-window-infinite-loader";

import SpinnerComponent from "@components/ui/spinner";
import useResize from "@hooks/useResize";
import ListRow from "@modules/virtual-list/row";
import { PHOTOS_LIMIT } from "@utils/constants";

import "./list.scss";

interface IVirtualGrid<T> {
	virtualRef: Ref<VirtualListHandle | null>;
	items: T[];
	options: {
		cols: number;
		gap: number;
		itemHeight: number;
	};
	hasMore: boolean;
	emptyText: string;
	isLoading: boolean;
	loadMore: (startIndex?: number, stopIndex?: number) => void;
	renderCb: (props: { item: T, itemWidth: number, itemHeight: number; }) => ReactNode;
	onScroll: (options: ListOnScrollProps) => void;
};

export interface VirtualListHandle {
	scrollTop(): void;
};

/**
 * Главный компонент виртуального списка.
 * В DOM отрисовываются только те элементы, которые находятся в видимой области.
 * Пакет "react-window-infinite-loader" автоматически управляет загрузкой элементов (повторные запросы при быстром скролле не идут). 
 */
export default forwardRef(function VirtualList<T extends { id: string; }>(
	props: IVirtualGrid<T>,
	parentRef: Ref<HTMLDivElement | null>,
) {
	const { virtualRef, items, options, hasMore, emptyText, isLoading, loadMore, renderCb, onScroll } = props;
	const { cols, gap, itemHeight } = options;

	const listRef = useRef<HTMLDivElement  | null>(null);

	const { width, height } = useResize(parentRef);
	// Реф для предыдущего значения items.length
	const prevItemsLength = useRef(items.length);

	/**
	 * Если мы подгрузили мало фотографий (вдруг сменили PHOTOS_LIMIT) или экран большой, то
	 * грузит до тех пор, пока не заполнил строками весь экран.
	 */
	useEffect(() => {
		// Сколько строк видно на экране
		const visibleRows = Math.floor(height / (itemHeight + gap));
		// Сколько строк у нас уже есть (без «строки загрузки»)
		const rowsLoaded = Math.ceil(items.length / cols);

		// Условие «недостаточно для экрана»:
		const needFillScreen = visibleRows >= rowsLoaded;
		// Условие «удалили что-то»:
		const didDelete = items.length < prevItemsLength.current;

		if ((needFillScreen || didDelete) && hasMore && !isLoading) {
			loadMore();
		}

		// Обновляем предыдущее значение items.length
		prevItemsLength.current = items.length;
	}, [ items.length, height, items.length, hasMore, isLoading ]);

	// Добавляем метод скролла виртуальному рефу, определенному в родителе (скролл по кнопке "Вверх")
	useImperativeHandle(virtualRef, () => ({
		scrollTop: () => {
			listRef.current?.scrollTo({
				top: 0,
				behavior: "smooth",
			});
		},
	}), []);

	// Динамический расчет общего количества строк
	const rowCount = useMemo(
		() => Math.ceil(items.length / cols) + (hasMore ? 1 : 0),
		[ items.length, hasMore ],
	);

	// Динамический расчет ширины картинки
	const itemWidth = useMemo(() => {
		return (width - (cols - 1) * gap) / cols;
	}, [ width ]);

	// Динамически проверяем есть ли уже этот ряд (загружен ли он)
	const isItemLoaded = useCallback((rowIndex: number) => {
		return rowIndex < Math.ceil(items.length / cols);
	}, [ items.length ]);

	// Генерируем уникальные ключи для каждой строки виртуального списка
	const getItemKey = useCallback((index: number) => {
		const first = items[index * cols];

		return first
			? `row-${first.id}`      // Ключ строки с фотографиями
			: `loading-${index}`;    // Ключ строки со спиннером загрузки
	}, [ items ]);

	/**
	 * Специально вынесено в useCallback, потому что react-window пересоздает каждый раз строку, 
	 * если указать ему для отрисовки анонимную функцию. Потому что она создается каждый раз при изменении элемента.
	 * В данном случае мы передаем в функцию отрисовки уже готовый объект, который не изменяется ни от каких параметров.
	 * Следовательно повторной отрисовки элемента при изменении высоты окна браузера не будет.
	 */
	const renderItem = useCallback(({ index, style, data }) => {
		return <ListRow<T>
			index={index}
			style={style}
			data={data}
			options={{ cols, gap, itemHeight }}
			renderCb={renderCb}
		/>;
	}, []);

	/**
	 * Показываем спиннер загрузки только в первый вход на страницу, последующие разы сбивают виртуальный список
	 * и он перерисовывается заново (скролл все время влзетает наверх).
	 */
	if (isLoading && !items.length) {
		return <div className="virtual-list__loading">
			<SpinnerComponent />
		</div>;
	}

	if (!items || !items.length) {
		return <div className="virtual-list__no-content opacity-text">
			{emptyText}
		</div>;
	}

	return <InfiniteLoader
		isItemLoaded={isItemLoaded}
		itemCount={rowCount}
		loadMoreItems={loadMore}
		threshold={2}
		minimumBatchSize={PHOTOS_LIMIT}
	>
		{({ onItemsRendered, ref: infiniteRef }) => {
			return <List
				ref={infiniteRef}
				outerRef={listRef}
				height={height}
				width={width}
				itemCount={rowCount}
				itemSize={itemHeight + gap}
				overscanCount={2}
				onItemsRendered={onItemsRendered}
				itemKey={getItemKey}
				itemData={{ items, itemWidth, hasMore }}
				onScroll={onScroll}
			>
				{renderItem}
			</List>;
		}}
	</InfiniteLoader>;
}) as <T extends { id: string; }>(props: IVirtualGrid<T> & { ref?: Ref<HTMLDivElement | null> }) => ReactElement;