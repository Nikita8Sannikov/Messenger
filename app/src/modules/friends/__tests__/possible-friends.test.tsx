import { FriendsTab } from "common-types";
import { createRef } from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import useUser from "@hooks/useUser";
import { type VirtualListHandle } from "@modules/virtual-list/list";
import i18next from "@service/i18n";
import useFriendsStore from "@store/friends";
import PossibleFriends from "../possible-friends";

jest.mock("@hooks/useUser");
jest.mock("@store/friends");
jest.mock("@service/i18n");

describe("PossibleFriends Component", () => {
	const mockVirtualRef = createRef<VirtualListHandle>();
	const mockFriendsService = {
		getFriends: jest.fn(),
		followFriend: jest.fn(),
	};

	beforeEach(() => {
		jest.clearAllMocks();

		(useUser as unknown as jest.Mock).mockReturnValue({
			friendsService: mockFriendsService,
		});

		(i18next.t as unknown as jest.Mock).mockImplementation((key: string) => key);

		const mockSetMainTab = jest.fn();
		const mockState = {
			searchFriends: [],
			isLoadingPossibleFriends: false,
			isLoadingFollowAction: false,
			setMainTab: mockSetMainTab,
		};

		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => selector(mockState));
		useFriendsStore.getState = () => ({
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			...mockState as any,
			setMainTab: mockSetMainTab,
		});
	});

	it("should render loading state correctly", () => {
		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => {
			const state = {
				searchFriends: [],
				isLoadingPossibleFriends: true,
				isLoadingFollowAction: false,
				setMainTab: jest.fn(),
			};
			return selector(state);
		});

		render(<PossibleFriends virtualRef={mockVirtualRef} />);
        
		expect(screen.getByTestId("possible-friends-container")).toBeInTheDocument();
		expect(screen.getByText("friends-module.possible_friends")).toBeInTheDocument();
	});

	it("should render empty state correctly", () => {
		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => {
			const state = {
				searchFriends: [],
				isLoadingPossibleFriends: false,
				isLoadingFollowAction: false,
				setMainTab: jest.fn(),
			};
			return selector(state);
		});

		render(<PossibleFriends virtualRef={mockVirtualRef} />);
        
		expect(screen.getByText("friends-module.no_other_users")).toBeInTheDocument();
	});

	it("should render list of possible friends correctly", () => {
		const mockFriends = [
			{
				id: "1",
				fullName: "John Doe",
				avatarUrl: "avatar1.jpg",
			},
			{
				id: "2",
				fullName: "Jane Smith",
				avatarUrl: "avatar2.jpg",
			},
		];

		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => {
			const state = {
				searchFriends: mockFriends,
				isLoadingPossibleFriends: false,
				isLoadingFollowAction: false,
				setMainTab: jest.fn(),
			};
			return selector(state);
		});

		render(<PossibleFriends virtualRef={mockVirtualRef} />);
        
		expect(screen.getByText("John Doe")).toBeInTheDocument();
		expect(screen.getByText("Jane Smith")).toBeInTheDocument();
		expect(screen.getAllByText("friends-module.add_to_friends")).toHaveLength(2);
	});

	it("should call followFriend when add button is clicked", async () => {
		const mockFriends = [ {
			id: "1",
			fullName: "John Doe",
			avatarUrl: "avatar1.jpg",
		} ];

		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => {
			const state = {
				searchFriends: mockFriends,
				isLoadingPossibleFriends: false,
				isLoadingFollowAction: false,
				setMainTab: jest.fn(),
			};
			return selector(state);
		});

		render(<PossibleFriends virtualRef={mockVirtualRef} />);
        
		const addButton = screen.getByText("friends-module.add_to_friends");
		fireEvent.click(addButton);

		await waitFor(() => {
			expect(mockFriendsService.followFriend).toHaveBeenCalledWith("1");
		});
	});

	it("should show loading state when follow action is in progress", () => {
		const mockFriends = [ {
			id: "1",
			fullName: "John Doe",
			avatarUrl: "avatar1.jpg",
		} ];

		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => {
			const state = {
				searchFriends: mockFriends,
				isLoadingPossibleFriends: false,
				isLoadingFollowAction: true,
				setMainTab: jest.fn(),
			};
			return selector(state);
		});

		render(<PossibleFriends virtualRef={mockVirtualRef} />);
        
		expect(screen.queryByText("friends-module.add_to_friends")).not.toBeInTheDocument();
		expect(screen.getByRole("progressbar")).toBeInTheDocument();
	});

	it("should call getFriends on mount", () => {
		render(<PossibleFriends virtualRef={mockVirtualRef} />);
        
		expect(mockFriendsService.getFriends).toHaveBeenCalledWith(FriendsTab.SEARCH);
	});

	it("should change main tab when title is clicked", () => {
		const mockSetMainTab = jest.fn();
		const mockState = {
			searchFriends: [],
			isLoadingPossibleFriends: false,
			isLoadingFollowAction: false,
			setMainTab: mockSetMainTab,
		};

		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => selector(mockState));
		useFriendsStore.getState = () => ({
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			...mockState as any,
			setMainTab: mockSetMainTab,
		});

		render(<PossibleFriends virtualRef={mockVirtualRef} />);
        
		const title = screen.getByText("friends-module.possible_friends");
		fireEvent.click(title);

		expect(mockSetMainTab).toHaveBeenCalledWith(FriendsTab.SEARCH);
	});

	it("should match snapshot", () => {
		const mockFriends = [ {
			id: "1",
			fullName: "John Doe",
			avatarUrl: "avatar1.jpg",
		} ];

		(useFriendsStore as unknown as jest.Mock).mockImplementation((selector) => {
			const state = {
				searchFriends: mockFriends,
				isLoadingPossibleFriends: false,
				isLoadingFollowAction: false,
				setMainTab: jest.fn(),
			};
			return selector(state);
		});

		const { container } = render(<PossibleFriends virtualRef={mockVirtualRef} />);
		expect(container).toMatchSnapshot();
	});
});
