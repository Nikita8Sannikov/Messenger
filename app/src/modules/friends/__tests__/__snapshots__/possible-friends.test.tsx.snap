// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PossibleFriends Component should match snapshot 1`] = `
<div>
  <div
    class="block-title"
  >
    friends-module.possible_friends
  </div>
  <div
    class="possible-container__block"
    data-testid="possible-friends-container"
  >
    <ul
      class="MuiList-root MuiList-padding possible-container__list css-b7dj7n-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters possible-container__item css-19f6boz-MuiListItem-root"
      >
        <div
          class="MuiListItemAvatar-root possible-container__item-avatar css-1ulznza-MuiListItemAvatar-root"
        >
          <span
            class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular avatar user-avatar big-avatar css-fmeqll-MuiAvatar-root"
              id="user-avatar"
            >
              <img
                alt="John Doe"
                class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                src="avatar1.jpg"
              />
            </div>
            <span
              class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
            />
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-multiline possible-container__item-block css-14ln1j6-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
          >
            John Doe
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
          >
            <span
              class="possible-container__action"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium possible-container__action-icon css-1umw9bq-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
                />
              </svg>
              <span
                class="possible-container__action-text"
              >
                friends-module.add_to_friends
              </span>
            </span>
          </p>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
